transport.go

```diff
@@ -202,6 +224,56 @@ func (t *transport) Dial(ctx context.Context, raddr ma.Multiaddr, p peer.ID) (tp
 	return conn, nil
 }
 
+func (t *transport) holePunch(ctx context.Context, p peer.ID, network string, addr *net.UDPAddr) error {
+	pconn, err := t.connManager.Dial(network, addr)
+	if err != nil {
+		return err
+	}
+	defer pconn.DecreaseCount()
+
+	ctx, cancel := context.WithTimeout(ctx, time.Second)
```

Comment:
Nit: put the `defer cancel()` here, not in L247.

Reply:
we can't really do that, it might have been cancelled by Listen in the meantime and we'll call it twice and panic.

Reply:
Note that we might want to add a limit to the number of packets (not just the timeout), so we can have two exit paths from the loop.

Reply:
You can call a `context.CancelFunc` as often as you want: https://golang.org/pkg/context/#CancelFunc

Reply:
oh interesting, I was under the strong impression that it would panic if you called it twice.

Reply:
ok, done.